portfolio/src/main/java/com/google/sps/servlets/CommentServlet.java

```diff
@@ -23,30 +29,47 @@
 public class CommentServlet extends HttpServlet {
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    Entity commentEntity = new Entity("Comment");
-    commentEntity.setProperty("username", getUsername(request));
-    commentEntity.setProperty("commentText", getComment(request));
-    commentEntity.setProperty("timestamp", System.currentTimeMillis());
+    UserService userService = UserServiceFactory.getUserService();
+    PrintWriter out = response.getWriter();
+
+    if (!userService.isUserLoggedIn()) {
+      out.print("<p>Please login before posting a comment </p>");
+      response.sendRedirect("/login");
+      return;
+    }
+
+    Entity commentEntity = new Entity(COMMENT_ENTITY);
+    commentEntity.setProperty(USERNAME_ENTITY_PROPERTY, getUsername(userService.getCurrentUser().getUserId()));
+    commentEntity.setProperty(COMMENT_TEXT_ENTITY_PROPERTY, request.getParameter(COMMENT_TEXT_ENTITY_PROPERTY));
+    commentEntity.setProperty(TIMESTAMP_ENTITY_PROPERTY, System.currentTimeMillis());
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
 
     datastore.put(commentEntity);
 
     response.sendRedirect("/community.html");
   }
 
-  private String getUsername(HttpServletRequest request) {
-    return request.getParameter("username");
+    private String getUsername(String id) {
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Query query = new Query(USER_ENTITY).setFilter(
+        new Query.FilterPredicate(ID_ENTITY_PROPERTY, Query.FilterOperator.EQUAL, id));
+    PreparedQuery results = datastore.prepare(query);
+    Entity entity = results.asSingleEntity();
+
+    return entity == null ? getRandomUsername() : (String) entity.getProperty(USERNAME_ENTITY_PROPERTY);
   }
 
-  private String getComment(HttpServletRequest request) {
-    return request.getParameter("comment");
+  private String getRandomUsername() {
+      String[] adjectives = {"Fantastic", "Random", "Honorable", "Jaded", "Fake", "Haunted", "Fighting", "Nappy", "Feirce", "Jealous", "Shark"};
+      String[] nouns = {"Goldfish", "Keyboard", "Pollution", "Bathroom", "Vehicle", "Week", "Republic", "Knowledge", "Economics"};
+      return adjectives[(int)(Math.random()*adjectives.length)] + "_" + nouns[(int)(Math.random()*nouns.length)];
```

Comment:
Suggest instead of plain arrays using Lists. Also Random class. Make sure to run the formatter.

```suggestion
      List<String> adjectives = Arrays.asList("Fantastic", "Random", "Honorable", "Jaded", "Fake", "Haunted", "Fighting", "Nappy", "Fierce", "Jealous", "Shark");
      List<String> nouns = Arrays.asList("Goldfish", "Keyboard", "Pollution", "Bathroom", "Vehicle", "Week", "Republic", "Knowledge", "Economics");
      
      Random random = new Random();
      return adjectives.get(random.nextInt(adjectives.size()) + "_" + nouns.get(random.nextInt(nouns.size());
```

Reply:
Randomed / listed